cpp/src/streaming/cudf/shuffler.cpp

```diff
@@ -10,37 +10,147 @@
 #include <rmm/cuda_stream_view.hpp>
 
 #include <rapidsmpf/cuda_event.hpp>
+#include <rapidsmpf/cuda_stream.hpp>
 #include <rapidsmpf/streaming/cudf/shuffler.hpp>
 
-namespace rapidsmpf::streaming::node {
+namespace rapidsmpf::streaming {
 
 namespace {
 
 /**
- * @brief Make @p primary wait until all work currently enqueued on @p secondary
- * completes.
+ * @brief Caller side corouting that inserts a partition ID into a set and notifies all
+ * waiting tasks.
  *
- * Records @p event on @p secondary and inserts a wait for that event on @p primary.
- * This is fully asynchronous with respect to the host thread; no host-side blocking.
+ * Wrapping mtx lock and cv notify in a coroutine task to avoid calling coro::sync_wait
+ * multiple times.
  *
- * @param primary The stream that must not run ahead.
- * @param secondary The stream whose already-enqueued work must complete first.
- * @param event The CUDA event to use for synchronization. The same event may be reused
- * across multiple calls; the caller does not need to provide an unique event each time.
+ * @param mtx The mutex to use for synchronization.
+ * @param cv The condition variable to use for notification.
+ * @param set The set to insert the partition ID into.
+ * @param pid The partition ID to insert.
+ * @return A coroutine task that completes when the partition ID is inserted into the set.
  */
-void sync_streams(
-    rmm::cuda_stream_view primary,
-    rmm::cuda_stream_view secondary,
-    cudaEvent_t const& event
+coro::task<void> insert_and_notify(
+    coro::mutex& mtx,
+    coro::condition_variable& cv,
+    std::unordered_set<shuffler::PartID>& set,
+    shuffler::PartID pid
 ) {
-    if (primary.value() != secondary.value()) {
-        RAPIDSMPF_CUDA_TRY(cudaEventRecord(event, secondary));
-        RAPIDSMPF_CUDA_TRY(cudaStreamWaitEvent(primary, event));
+    // Note: this coroutine is not needed to be scheduled, because it is called from the
+    // progress thread.
+    {
+        auto lock = co_await mtx.scoped_lock();
+        set.insert(pid);
     }
+    co_await cv.notify_all();
 }
 
 }  // namespace
 
+ShufflerAsync::ShufflerAsync(
+    std::shared_ptr<Context> ctx,
+    rmm::cuda_stream_view stream,
+    OpID op_id,
+    shuffler::PartID total_num_partitions,
+    shuffler::Shuffler::PartitionOwner partition_owner
+)
+    : ctx_(std::move(ctx)),
+      shuffler_(
+          ctx_->comm(),
+          ctx_->progress_thread(),
+          op_id,
+          total_num_partitions,
+          stream,
+          ctx_->br(),
+          [this](shuffler::PartID pid) -> void {
+              ctx_->comm()->logger().trace("inserting finished partition ", pid);
```

Comment:
*nit*: This message is slightly misleading, since we might expect that to mean that `insert_finished` has been called with this pid.

How about something like `"notifying waiters that ", pid, " is ready"`?